### DISCUSSION TITLE
Cross-Imports

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#390] [DocDraft]
Про кросс-импорты в слоях

**Comment by azinit** [comment]
https://t.me/feature_sliced/9171

**Comment by azinit** [comment]
Про кросс-импорты фич и сущностей

https://t.me/feature_sliced/11378

**Comment by illright** [comment]
## RFC 1. Explicit index files for every pair of slices that need cross-importing

> **Note**
> We're only talking about cross-imports between slices on `entities` for now, all of the layers above still have cross-imports forbidden.

I suggest to use the following convention:

<table>
<tr><td>
<img width="286" alt="image" src="https://user-images.githubusercontent.com/15035286/230884436-6093bf23-0aaf-4ad3-9f81-94fdc1d86f15.png">
</td>
<td>

```ts
// File: entities/city/api/useCities.ts

import { ONLINE_FORMAT } from "entities/country/@x/city";
import { supabase } from "shared/api";
import type { Option } from "shared/ui";
```

</td></tr></table>

### Advantages over plain cross-importing

1. Writing such index files is slightly cumbersome, which discourages cross-importing and encourages finding alternative solutions that require no cross-imports
2. The list of types and objects that are shared between slices is explicit and predictable, eliminating the step of finding all objects that are cross-imported during refactors
3. Each slice has its own index file, therefore it's easy to find all of the dependents of a particular slice and refactor them individually. When you want to remove an object that has been referenced in another slice, you can temporarily move that object into the index file to avoid breaking the dependency, and refactor it away whenever it's convenient
4. The module structure is standardized, meaning it would be simpler to create tools to lint such cross-imports
5. It reads nice in the import string: "import object from (entity country)-x-city"

### Variations

The name and the special symbol of the special segment are subject to change. The special symbol was introduced to minimize collisions with existing segments as well as to delineate this as a more advanced concept.

This convention was inspired by [meta-categories (in Russian)](https://github.com/austrokhart/virtual-nesting#%D0%BC%D0%B5%D1%82%D0%B0-%D0%BA%D0%B0%D1%82%D0%B5%D0%B3%D0%BE%D1%80%D0%B8%D0%B8) in the Virtual Nesting methodology.

### How do we teach this?

I think it's best not to mention this at first when explaining the principles of FSD. The explanation of layers will not change, except, maybe, for a side note that links to a more advanced article. That article will introduce this convention. We should also add an entry to the FAQ that will link to that article.

That way, we avoid creating too many exceptions for our rules and lead beginners to the more difficult yet more rewarding path of separating domains (encouraging them to _think_).

**Reply by noveogroup-amorgunov** [reply]
Great proposal. I tried to use it in my project, all good.

<img width="996" alt="Снимок экрана 2023-04-23 в 23 49 19" src="https://user-images.githubusercontent.com/21186268/233865577-b25a318c-1b23-441b-b4e1-19f89e0be964.png">


I updated eslint-rule (@feature-sliced/eslint-config) for it (https://github.com/noveogroup-amorgunov/nukeapp/blob/v0.0.1/patches/%40feature-sliced%2Beslint-config%2B0.1.0-beta.6.patch), you can use it if RFC will added to methodology
